Cria README do projeto (#1)
betrybe/sd-0x-inventory-report: README.md

- Através da importação de um arquivo `JSON`;

- Através da importação de um arquivo `XML`;

Além disso, o relatório final deverá ser gerável em duas versões: simples e completa.

### Como o projeto deve ser executável

Seu programa deverá ser executável **via linha de comando** com o comando `python3 main.py argumento1 argumento2`:

- O **argumento 1** deve receber o caminho de um arquivo a ser importado. O arquivo pode ser um `csv`, `json` ou `xml`.

- O **argumento 2** pode receber duas strings: `simples` e `completo`, cada uma gerando o respectivo relatório.

---

## Desenvolvimento e testes

Este repositório já contém um _template_ com a estrutura de diretórios e arquivos, tanto de código quanto de teste criados. Veja abaixo:

```
.
├── README.md
├── setup.cfg
├── requirements.txt
├── data
│   ├── inventory_20200823.csv
│   ├── inventory_20200823.json
│   └── inventory_20200823.xml
├── main.py
├── inventory
│   ├── inventory.py
│   └── inventory_iterator.py
├── importer
│   ├── importer.py
│   ├── csv_importer.py
│   ├── json_importer.py
│   └── xml_importer.py
├── reports
│   ├── simple_report.py
│   └── complete_report.py
├── tests

> │   ├── main.py

Reviewer: Os arquivos de teste deveriam começar com o prefixo `test_` ou terminar com o sufixo `_test`. Tipo:

```suggestion
│   ├── main_test.py
```

    "nome_da_empresa":"Forces of Nature",
    "data_de_fabricacao":"2020-07-04",
    "data_de_validade":"2023-02-09",
    "numero_de_serie":"FR48 2002 7680 97V4 W6FO LEBT 081",
    "instrucoes_de_armazenamento":"in blandit ultrices enim lorem ipsum dolor sit amet consectetuer adipiscing elit proin interdum mauris non ligula pellentesque ultrices phasellus"
  }
]
```

### Importação de arquivos XML

Os arquivos **XML** seguem o seguinte modelo:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<dataset>
  <record>
    <id>1</id>
    <nome_do_produto>valsartan and hydrochlorothiazide</nome_do_produto>
    <nome_da_empresa>Lake Erie Medical &amp; Surgical Supply DBA Quality Care Products LLC</nome_da_empresa>
    <data_de_fabricacao>2019-10-27</data_de_fabricacao>
    <data_de_validade>2022-08-31</data_de_validade>
    <numero_de_serie>MT08 VVDN 2131 9NFL C1JG KTDV RS1L LOZ</numero_de_serie>
    <instrucoes_de_armazenamento>at lorem integer tincidunt ante vel ipsum praesent blandit lacinia erat</instrucoes_de_armazenamento>
  </record>
</dataset>
```

---

## Requisitos obrigatórios:

> #### 1 - Deve haver uma função `generate` numa classe `SimpleReport` do módulo `simple-report`. Essa função deverá receber dados numa estrutura `dict` e deverá gerar uma saída para a linha de comando.

Reviewer: ```suggestion
#### 1 - Deve haver uma função `generate` numa classe `SimpleReport` do módulo `simple-report`. Essa função deverá receber dados numa lista contendo estruturas do tipo `dict` e deverá gerar uma saída para a linha de comando.
```

##### As seguintes verificações serão feitas:

> - A função deve receber de parâmetro um dict no seguinte formato:

Reviewer: ```suggestion
- A função deve receber de parâmetro uma lista de dicionários no seguinte formato:
```

```python
[
  {
    "id": 1,
    "nome_do_produto": "CALENDULA OFFICINALIS FLOWERING TOP, GERANIUM MACULATUM ROOT, SODIUM CHLORIDE, THUJA OCCIDENTALIS LEAFY TWIG, ZINC, and ECHINACEA ANGUSTIFOLIA",
    "nome_da_empresa": "Forces of Nature",
    "data_de_fabricacao": "2020-07-04",
    "data_de_validade": "2023-02-09",
    "numero_de_serie": "FR48 2002 7680 97V4 W6FO LEBT 081",
    "instrucoes_de_armazenamento": "in blandit ultrices enim lorem ipsum dolor sit amet consectetuer adipiscing elit proin interdum mauris non ligula pellentesque ultrices phasellus"
  }
]
```

- A função deverá gerar, na linha de comando, uma saída com o seguinte formato:

```bash
Data de fabricação mais antiga: YYYY-MM-DD
Data de validade mais próxima: YYYY-MM-DD
Empresa com maior quantidade de produtos estocados: NOME DA EMPRESA
```

**Dica**: O módulo [datetime](https://docs.python.org/3/library/datetime.html) vai te ajudar.

> #### 2 - Deve haver uma função `generate` numa classe `CompleteReport` do módulo `complete-report`. Essa função deverá receber dados numa estrutura `dict` e deverá gerar uma saída para a linha de comando.

Reviewer: ```suggestion
#### 2 - Deve haver uma função `generate` numa classe `CompleteReport` do módulo `complete-report`. Essa função deverá receber dados numa lista contendo estruturas do tipo `dict` e deverá gerar uma saída para a linha de comando.
```

##### As seguintes verificações serão feitas:

- A função deve ser herdeira da classe `SimpleReport`, de modo a especializar seu comportamento.

> - A função deve receber de parâmetro um dict no seguinte formato:

Reviewer: ```suggestion
- A função deve receber de parâmetro uma lista de dicionários no seguinte formato:
```

#### 4 - Deve haver uma função `import` dentro de uma classe `Inventory` do módulo `inventory`, capaz de ler um arquivo JSON passado como parâmetro de linha de comando

##### As seguintes verificações serão feitas:

- A função, quando receber um arquivo JSON, deve chamar a função de gerar relatório correspondente à entrada passada, `simples` ou `completo`, com os dados importados corretamente passados por parâmetro a elas.

#### 5 - Deve haver uma função `import` dentro de uma classe `Inventory` do módulo `inventory`, capaz de ler um arquivo XML passado como parâmetro  de linha de comando

##### As seguintes verificações serão feitas:

- A função, quando receber um arquivo XML, deve chamar a função de gerar relatório correspondente à entrada passada, `simples` ou `completo`, com os dados importados corretamente passados por parâmetro a elas.

#### 6 - Deve haver uma classe abstrata `Importer` no módulo import. Deve haver três classes herdeiras desta: `CsvImporter`, `JsonImporter` e `XmlImporter`, cada uma definida em seu respectivo módulo.

##### As seguintes verificações serão feitas:

> - A classe abstrata deve definir um método de cabeçalho (uma interface) `import` a ser implementado por cada classe herdeira. Ela deve receber como parâmetro o nome do arquivo a ser importado.

Reviewer: O que você quis dizer por "método de cabeçalho"? Isso me deixou um pouco confuso em relação ao intuito dessa verificação.

Author: É uma interface mesmo. A classe abstrata define uma função que as herdeiras tem que implementar. Como vc acha que fica melhor descrito?

Reviewer: Saquei. Se é esse o caso, creio que você poderia usar o termo método abstrato mesmo, ou pedir pra definir a assinatura do método. O que achar que fica mais intuitivo.

> - A função import definida por cada classe herdeira deve lançar uma exceção caso a extensão do arquivo passado por parâmetro seja inválida (por exemplo, quando se passa um `csv` para o `JsonImporter`).

Reviewer: ```suggestion
- A função `import` definida por cada classe herdeira deve lançar uma exceção caso a extensão do arquivo passado por parâmetro seja inválida (por exemplo, quando se passa um CSV para o `JsonImporter`).
```

> - A função deverá ler os dados do arquivo passado e retorná-los estruturados em um `dict` conforme exemplo abaixo:

Reviewer: ```suggestion
- A função deverá ler os dados do arquivo passado e retorná-los estruturados em uma lista de dicionários conforme exemplo abaixo:
```

```python
[
  {
    "id": 1,
    "nome_do_produto": "CALENDULA OFFICINALIS FLOWERING TOP, GERANIUM MACULATUM ROOT, SODIUM CHLORIDE, THUJA OCCIDENTALIS LEAFY TWIG, ZINC, and ECHINACEA ANGUSTIFOLIA",
    "nome_da_empresa": "Forces of Nature",
    "data_de_fabricacao": "2020-07-04",
    "data_de_validade": "2023-02-09",
    "numero_de_serie": "FR48 2002 7680 97V4 W6FO LEBT 081",
    "instrucoes_de_armazenamento": "in blandit ultrices enim lorem ipsum dolor sit amet consectetuer adipiscing elit proin interdum mauris non ligula pellentesque ultrices phasellus"
  }
]
```

- A classe `Inventory` deve utilizar as classes definidas neste requisito para lidar com a lógica de importação, via **composição**.

#### 7 - Deve haver uma classe `InventoryIterator` no módulo `inventory-iterator`, que implementa a interface de um iterator. A classe `Inventory` deve implementar o método `__iter__` associado a essa classe.

##### As seguintes verificações serão feitas:

> - As classes `InventoryIterator` e `Inventory` devem implementar corretamente a interface de um iterator, de modo que o código abaixo nos dê o primeiro item do dicionário de dados importados:

Reviewer: ```suggestion
- As classes `InventoryIterator` e `Inventory` devem implementar corretamente a interface de um iterator, de modo que o código abaixo nos dê o primeiro item da lista de dicionários com os dados importados:
```

```python
# ... Acima, um código que instancia e importa um arquivo para a variável `inventory` e importações do módulo Iterator e Iterable

iterator = iter(inventory)
first_item = next(iterator)
```

## Requisitos bônus:

#### 8 - A cobertura de testes do projeto deve ser de no mínimo 90%.

##### As seguintes verificações serão feitas:

- Todos os testes que envolvem mensagens na saída padrão ou de erro, devem ter sua saída redirecionada para Fakes com `StringIO`;

- Todos os testes que envolvem manipulação de arquivos criam Fakes com `StringIO`;

> - Todas as requisições externas a uma classe utilizam _Mocks_;

Reviewer: Nesse projeto não temos requisições externas não, temos? Pelo menos creio não ter visto nenhuma nos requisitos.

Author: Eu pensei em `externas` como requisições dessa classe pra outras classes. Mockar tudo pra ficar bem unitário. Acha que faz sentido? Certamente vale esclarecer.

Reviewer: Faz sentido, afinal pra testes de integração você usaria mocks. Só que do jeito que isso está escrito, pelo menos para mim, não ficou claro que era esse o intuito.

Talvez valha a pena ser mais específico aqui para dar a entender isso, não?
